### PR TITLE
feat: support new object upload

### PR DESCRIPTION
**Pull request by ryanswrt**
This PR supports the new object upload format and does automatic upload when objects are passed to generated functions

### CODE REVIEW

File: nbs/Tutorials/gmx_restart.ipynb [hunk 1]
```diff
@@ -45,7 +47,7 @@
      "name": "stdout",
      "output_type": "stream",
      "text": [
-      "2024-03-22 16:01:05,862 - rush - INFO - Restoring by default via env\n"
+      "2024-04-10 17:21:43,192 - rush - INFO - Not restoring by default via default\n"
```
Comment (s-i-l-k-e) [inline nbs/Tutorials/gmx_restart.ipynb:50]: This seems bad

File: nbs/Tutorials/gmx_restart.ipynb [hunk 2]
```diff
@@ -74,7 +76,44 @@
      "name": "stdout",
      "output_type": "stream",
      "text": [
-      "2024-03-22 16:01:05,887 - rush - INFO - Restoring by default via env\n"
+      "2024-04-10 17:21:43,456 - rush - INFO - Not restoring by default via default\n",
```
Comment (s-i-l-k-e) [inline nbs/Tutorials/gmx_restart.ipynb:79]: Not sure we want this output to appear on the docs site

File: nbs/Tutorials/gmx_restart.ipynb (hunk 1 above)
Comment (s-i-l-k-e) [reply nbs/Tutorials/gmx_restart.ipynb:50]: It's a really confusing phrase; I think it could be improved.

Comment (s-i-l-k-e) [reply nbs/Tutorials/gmx_restart.ipynb:50]: Don't think it was introduced in this PR, but nonetheless good to fix

Comment (ryanswrt) [reply nbs/Tutorials/gmx_restart.ipynb:50]: The notebooks shouldn't restore in the documentation context

File: rush/async_utils.py [hunk 3]
```diff
@@ -23,10 +23,20 @@ def asyncio_run(coro: Coroutine[Any, Any, T]) -> T:
     :param coro: A coroutine, typically an async method
     :param timeout: How many seconds we should wait for a result before raising an error
     """
+    try:
+        # FIXME: this is a hack to work around an annoying stall issue in Jupyter notebooks
```
Comment (s-i-l-k-e) [inline rush/async_utils.py:27]: Would be good to provide a bit more detail here

Comment (s-i-l-k-e) [reply rush/async_utils.py:27]: Not a major issue if it's on your TODO list though

File: rush/graphql_client/client.py [hunk 4]
```diff
@@ -809,293 +667,37 @@ async def module_instance_minimal(self, id: UUID, **kwargs: Any) -> ModuleInstan
         data = self.get_data(response)
         return ModuleInstanceMinimal.model_validate(data).module_instance
 
-    async def object_url(self, id: UUID, **kwargs: Any) -> Optional[ObjectUrlObject]:
+    async def object_url(self, path: Any, **kwargs: Any) -> ObjectUrlObjectPath:
```
Comment (s-i-l-k-e) [inline rush/graphql_client/client.py:670]: Why did you need to make `path: Any`

File: rush/provider.py [hunk 5]
```diff
@@ -58,17 +58,23 @@
...
-from .typedef import SCALARS, build_typechecker, type_from_typedef
+from .typedef import SCALARS, RushType, build_typechecker, type_from_typedef
 
 if sys.version_info >= (3, 12):
     from .types import ArgId, ModuleInstanceId, Resources, Target
 else:
     from .legacy_types import ArgId, ModuleInstanceId, Resources, Target
 
 
+class RushVirtualObject:
```
Comment (s-i-l-k-e) [inline rush/provider.py]: Why do you need to prefix this class with `Rush`?

File: rush/provider.py [hunk 6]
```diff
@@ -1073,18 +1098,22 @@ async def retry(
...
+        if isinstance(file, str):
+            file = Path(file)
         with open(file, "rb") as f:
-            return await self.client.upload_arg(
+            format = ObjectFormat.JSON if file.suffix == ".json" else ObjectFormat.BIN
+            return await self.client.upload_object(
                 typeinfo=typeinfo,
-                file=Upload(filename=f.name, content=f, content_type="application/octet-stream"),
+                format=format,
+                file=Upload(filename=f.name, content=f, content_type="application/text"),
```
Comment (s-i-l-k-e) [inline rush/provider.py]: I don't actually think this is a valid MIME-type

File: rush/typedef.py [hunk 7]
```diff
@@ -256,7 +258,12 @@ def to_python_type(self) -> type[RushObject[Any]]:
         return RushObject[self.t.to_python_type()]
 
     def matches(self, other: Path | StringIO | BytesIO | Any) -> tuple[bool, str | None]:
-        if isinstance(other, _RushObject) or isinstance(other, Path) or isinstance(other, StringIO):
+        if (
+            isinstance(other, _RushObject)
```
Comment (s-i-l-k-e) [inline rush/typedef.py]: `isinstance` can take a tuple of types, and it would be a cleaner way to express this rather than a number of chained `or`s

**Review by s-i-l-k-e — APPROVED** [review]
